src/fparser/two/Fortran2003.py

```diff
@@ -1554,179 +1589,341 @@ def match(string):
+    ''':F03R:`451`::
+
+    Fortran 2003 rule R451
+    that specifies syntax of specific binding for a type-bound
+    procedure within a derived type.
+
     <specific-binding> = PROCEDURE [ ( <interface-name> ) ] [
         [ , <binding-attr-list> ] :: ] <binding-name> [ => <procedure-name> ]
-    """
+
+    The following are associated constraints:
+
+    "C456 (R451) If => procedure-name appears, the double-colon
+    separator shall appear."
+
+    "C457 (R451) If => procedure-name appears, interface-name shall not
+    appear."
+
+    "C458 (R451) The procedure-name shall be the name of an accessible
+    module procedure or an external procedure that has an explicit
+    interface." Note, this is not checked by fparser.
+
+    '''
     subclass_names = []
     use_names = ['Interface_Name', 'Binding_Attr_List',
                  'Binding_Name', 'Procedure_Name']
 
     @staticmethod
     def match(string):
-        if string[:9].upper() != 'PROCEDURE':
+        '''
+        :param str string: Fortran code to check for a match
+        :return: 5-tuple containing strings and instances of the classes
+                 describing a specific type-bound procedure (optional
+                 interface name, optional binding attribute list,
+                 optional double colon delimiter, mandatory binding
+                 name and optional procedure name)
+        :rtype: 5-tuple of objects (1 mandatory and 4 optional)
+        '''
+        # Remove any leading, trailing spaces.
+        string_strip = string.strip()
+        if string_strip[:9].upper() != 'PROCEDURE':
+            # There is no 'PROCEDURE' statement.
             return
-        line = string[9:].lstrip()
+        line = string_strip[9:].lstrip()
+        # Find optional interface name if it exists.
         iname = None
         if line.startswith('('):
-            i = line.find(')')
-            if i == -1:
-                return
-            iname = Interface_Name(line[1:i].strip())
-            line = line[i+1:].lstrip()
+            index = line.find(')')
+            if index != -1:
+                iname = Interface_Name(line[1:index].strip())
+                line = line[index+1:].lstrip()
+        # Look for optional double colon and binding attribute list.
+        dcolon = None
         mylist = None
-        i = line.find('::')
-        if i != -1:
+        index = line.find('::')
+        if index != -1:
+            dcolon = '::'
             if line.startswith(','):
-                mylist = Binding_Attr_List(line[1:i].strip())
-            line = line[i+2:].lstrip()
-        i = line.find('=>')
+                mylist = Binding_Attr_List(line[1:index].strip())
+            elif line[:index].strip():
+                # There is content between procedure (with optional
+                # interface) and :: that does not start with a ','
+                # which is a syntax error.
+                return
```

Comment:
We're certain that there is no match because there would be a syntax error at this point, however we are not certain that the code in question will not match another set of rules, so we can not assume that this actually is a syntax error.

It may be that at this point no other rule could have matched but we would need to look carefully. I think a better solution for the future would be to return a no-match-error when we do not match and for these errors to be examined to see if there is only one or a small number, and they could be returned. I've added this suggestion to the developers documentation in one of the issues/PR's but I'm not sure which one!